python_modules/dagster/dagster/_experimental/staging_io_manager.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any, Optional, Sequence, Union
+
+from dagster import InputContext, OutputContext
+from dagster._core.definitions.events import AssetMaterialization, AssetObservation
+from dagster._core.definitions.metadata import (
+    MetadataEntry,
+    PartitionMetadataEntry,
+    TextMetadataValue,
+)
+from dagster._core.events import AssetObservationData, DagsterEventType, StepMaterializationData
+from dagster._core.events.log import EventLogEntry
+from dagster._core.storage.io_manager import IOManager
+
+
+def get_asset_materialization_from_event_log_entry(
+    event_log_entry: EventLogEntry,
+) -> Optional[AssetMaterialization]:
+    if event_log_entry.dagster_event_type != DagsterEventType.ASSET_MATERIALIZATION:
+        return None
+    dagster_event = event_log_entry.dagster_event
+    if not dagster_event:
+        return None
+    if not isinstance(dagster_event.event_specific_data, StepMaterializationData):
+        return None
+
+    return (
+        dagster_event.event_specific_data.materialization
+        if isinstance(dagster_event.event_specific_data.materialization, AssetMaterialization)
+        else None
+    )
+
+
+def get_asset_observation_from_event_log_entry(
+    event_log_entry: EventLogEntry,
+) -> Optional[AssetObservation]:
+    if event_log_entry.dagster_event_type != DagsterEventType.ASSET_OBSERVATION:
+        return None
+    dagster_event = event_log_entry.dagster_event
+    # assert dagster_event
+    if not dagster_event:
+        return None
+    # assert isinstance(dagster_event.event_specific_data, AssetObservationData)
+    if not isinstance(dagster_event.event_specific_data, AssetObservationData):
+        return None
+
+    return (
+        dagster_event.event_specific_data.asset_observation
+        if isinstance(dagster_event.event_specific_data.asset_observation, AssetObservation)
+        else None
+    )
+
+
+def get_text_metadata(
+    metadata_entries: Sequence[Union[MetadataEntry, PartitionMetadataEntry]], label
+) -> Optional[str]:
+    for metadata_entry in metadata_entries:
+        entry_to_search = (
+            metadata_entry if isinstance(metadata_entry, MetadataEntry) else metadata_entry.entry
+        )
+        if entry_to_search.label == label and isinstance(
+            entry_to_search.entry_data, TextMetadataValue
+        ):
+            return entry_to_search.entry_data.value
+
+    return None
+
+
+class DualEnvironmentIOManager(IOManager):
+    def __init__(
+        self, *, prod_io_manager, staging_io_manager, env_key="env", staging_label="staging"
```

Comment:
"env" is a very general term. Would bias towards trying to find something more specific - e.g. "storage_env".

Reply:
agree